### PR TITLE
Add support for dictionary with torch object keys.

### LINKED ISSUE #101979: [dynamo] Dynamo struggles with dictionary of dictionaries of torch objects
### 🐛 Describe the bug

The following example

```python
import torch


_result_type_dict = {
    torch.float16: {
        torch.float32: torch.float32,
    },
}

@torch.compile
def f():
    return torch.randn(3, dtype=_result_type_dict[torch.float16][torch.float32])

f()
```

errors out in the following way:

```python
Traceback (most recent call last):
  File "/home/lezcano/git/pytorch/pytorch/torch/_dynamo/convert_frame.py", line 470, in _compile
    check_fn = CheckFunctionManager(
  File "/home/lezcano/git/pytorch/pytorch/torch/_dynamo/guards.py", line 707, in __init__
    guard.create(local_builder, global_builder)
  File "/home/lezcano/git/pytorch/pytorch/torch/_guards.py", line 196, in create
    return self.create_fn(self.source.select(local_builder, global_builder), self)
  File "/home/lezcano/git/pytorch/pytorch/torch/_dynamo/guards.py", line 195, in TYPE_MATCH
    t = type(self.get(guard.name))
  File "/home/lezcano/git/pytorch/pytorch/torch/_dynamo/guards.py", line 171, in get
    return eval(name, self.scope, CLOSURE_VARS)
  File "<string>", line 1, in <module>
NameError: name 'torch' is not defined
```

This was found when integrating of https://github.com/Quansight-Labs/numpy_pytorch_interop with `torch.compile`

### Versions

master

cc @ezyang @msaroufim @wconstab @ngimel @bdhirsh @anijain2305 @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @Xia-Weiwen @wenzhe-nrv @jiayisunx

### PR DESCRIPTION
**Pull request by ysiraichi**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #103158

Fixes: #101979

This PR adds support for dictionaries with torch object as keys in dynamo.

The main problem was that, for example, the source built for `d[torch.float]` (`d` being a
dictionary) was `ODictGetItemSource(GlobalSource('d'), index=torch.float)`. When
`Source.name` method was called, we got `odict_getitem(G['d'], torch.float)`. Evaluating
that string raised an error, since `torch` was only available in the global dictionary `G`
as `G["torch"]`.

Instead, this PR builds the source:
`ODictGetItemSource(GlobalSource('d'), index=AttrSource(GlobalSource('torch'), 'float'))`.
The to-be-evaluated string is correctly generated as:
`odict_getitem(G['d'], G['torch'].float)`.

Here's a minimal example that reproduces the error, before this PR:

```python
import torch

d = {
    torch.float16: torch.float32,
}

@torch.compile
def f():
    return torch.randn(3, dtype=d[torch.float16])

f()
```

cc @voznesenskym @penguinwu @anijain2305 @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @Xia-Weiwen @wenzhe-nrv @jiayisunx @ipiszy @aakhundov

### CODE REVIEW

File: torch/_dynamo/source.py [hunk 1]
```diff
@@ -399,10 +399,15 @@ def __post_init__(self):
         assert self.base is not None
 
     def reconstruct(self, codegen):
```
Comment (mlazos) [inline torch/_dynamo/source.py:401]: This is nearly identical to the standard `GetItemSource` above, can you abstract this out to a common place?

File: torch/_dynamo/variables/user_defined.py [hunk 2]
```diff
@@ -502,14 +502,13 @@ def call_hasattr(self, tx, name: str) -> "VariableTracker":
     def odict_getitem(self, tx, key):
         from .builder import VariableBuilder
 
+        index = key.source if key.source is not None else key.as_python_constant()
```
Comment (mlazos) [inline torch/_dynamo/variables/user_defined.py:505]: You should check directly if this is a supported type vs checking if the source is not `None`. A lot of types will meet this constraint

Comment (ysiraichi) [reply torch/_dynamo/variables/user_defined.py:505]: Should I do it like in [this comment](https://github.com/pytorch/pytorch/issues/101979#issuecomment-1574438751)?

File: torch/_dynamo/source.py (hunk 1 above)
Comment (ysiraichi) [reply torch/_dynamo/source.py:401]: As far as I understand, the identical parts of the 2 are:
- Reconstruct the `base`
- Check if the `index` is an instance of `Source`
- Maybe (check the differences) load the `index` if it is not 

The differences are:
- `GetItemSource` still has to check if `index` is a slice
- `GetItemSource` creates a `BINARY_SUBSCR` instruction
- `ODictGetItemSource` has to load the `getitem` function
- `ODictGetItemSource` has to call this function (another bytecode)

Do you think that I should do it, regardless of the differences?

Comment (mlazos) [reply torch/_dynamo/source.py:401]: If you abstract out a bytecode_suffix method that is overridden in each subclass, and say index_is_slice is always false for OdictGetItem, I think the differences resolve, right?

File: torch/_dynamo/variables/user_defined.py (hunk 2 above)
Comment (mlazos) [reply torch/_dynamo/variables/user_defined.py:505]: So I'm thinking we should have an `is_valid_keys` function which decides whether a type is permitted or not, this already exists for `ConstDictVariable`. Can you augment this to check if it is a valid `TorchVariable` representing a dtype and make sure both `ConstDictVariable` and ODicts use that function to check for valid keys?

File: torch/_dynamo/source.py (hunk 1 above)
Comment (ysiraichi) [reply torch/_dynamo/source.py:408]: Yes. You are totally right. Missed that completely.

File: torch/_dynamo/variables/user_defined.py [hunk 3]
```diff
@@ -502,14 +503,18 @@ def call_hasattr(self, tx, name: str) -> "VariableTracker":
     def odict_getitem(self, tx, key):
         from .builder import VariableBuilder
 
+        index = (
+            key.source
+            if ConstDictVariable.is_valid_key(key) and key.source is not None
+            else key.as_python_constant()
+        )
```
Comment (ysiraichi) [inline torch/_dynamo/variables/user_defined.py:510]: Does this look better?

Comment (mlazos) [reply torch/_dynamo/variables/user_defined.py:510]: Yup thank you!

**Review by mlazos — APPROVED** [review]

**Pull request closed without merging**